perception/navigator_vision/nodes/pc_colorizer.cpp

```diff
@@ -0,0 +1,122 @@
+#include <ros/ros.h>
+#include <deque>
+
+#include <sensor_msgs/PointCloud2.h>
+
+#include <mil_vision_lib/image_acquisition/ros_camera_stream.hpp>
+
+#include <geometry_msgs/TransformStamped.h>
+#include <tf2_ros/transform_listener.h>
+
+#include <pcl/PCLPointCloud2.h>
+#include <pcl/conversions.h>
+#include <pcl/point_types.h>
+#include <pcl_conversions/pcl_conversions.h>
+#include <pcl_ros/point_cloud.h>
+#include <pcl_ros/transforms.h>
+
+#include <opencv2/opencv.hpp>
+
+#include <Eigen/Dense>
+#include <Eigen/Geometry>
+
+using ROSCameraStream_Vec3 = mil_vision::ROSCameraStream<cv::Vec3b>;
+//! Point Cloud Colorizer class.
+/*! Subscribes to cameras and velodyne topic to publish a RGB pointcloud.
+  Inputs: rosparam '/pc_colorizer/cameras', and topic '/velodyne_points'
+  Outputs: topic '/velodyne_points/colored'
+*/
+class Colorizer
+{
+  ros::NodeHandle nh_;
+  //! Deque to store camera streams for each camera
+  std::deque<std::unique_ptr<ROSCameraStream_Vec3>> cameras_;
+
+  ros::Subscriber point_cloud_sub_;
+  ros::Publisher rgb_cloud_pub_;
+
+  tf2_ros::Buffer tf_buffer_;
+  tf2_ros::TransformListener tf_listener_;
+
+public:
+  //! Initialize tf buffer and resolve camera feeds
+  Colorizer() : nh_(), tf_listener_(tf_buffer_)
+  {
+    std::vector<std::string> a = { "/stereo/right/image_rect_color", "/camera/front/right/image_rect_color" };
+    nh_.param<std::vector<std::string>>("/pc_colorizer/cameras", a, a);
+    // pushback camera streams that are avaliable
+    for (auto &i : a)
+    {
+      std::unique_ptr<ROSCameraStream_Vec3> cam(new ROSCameraStream_Vec3(nh_, 1));
+      cam->init(i);
+      if (cam->ok())
+        cameras_.push_back(std::move(cam));
+    }
+    point_cloud_sub_ =
+        nh_.subscribe<pcl::PointCloud<pcl::PointXYZRGB>>("/velodyne_points", 1, &Colorizer::cb_velodyne, this);
+    rgb_cloud_pub_ = nh_.advertise<pcl::PointCloud<pcl::PointXYZRGB>>("/velodyne_points/colored", 1);
+  }
+
+  void cb_velodyne(const pcl::PointCloud<pcl::PointXYZRGB>::ConstPtr &point_cloud)
+  {
+    pcl::PointCloud<pcl::PointXYZRGB>::Ptr temp_cloud(new pcl::PointCloud<pcl::PointXYZRGB>(*point_cloud));
+
+    for (pcl::PointXYZRGB &point : temp_cloud->points)
+    {
+      for (auto &&cam : cameras_)
+      {
+        if (cam->size() <= 0)
+          break;
+        if (!tf_buffer_.canTransform((*cam)[0]->getCameraModelPtr()->tfFrame(), "velodyne", ros::Time(0)))
```

Comment:
Instead of hardcoding "velodyne" frame, use frame from message header so it is more generic